### PR TITLE
add docker build to actions pipeline

### PR DESCRIPTION
**Pull request by nicolasochem**
This builds a breadcrumbs container and pushes it to the github registry for every branch, PR or release.

It is useful to me since I am running breadcrumbs in a k8s cluster.

See example: https://github.com/nicolasochem/breadcrumbs/pkgs/container/breadcrumbs

### CODE REVIEW

**Comment by nicolasochem** [comment]
This also requires changes in the github organization to allow pushes by this repo.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,13 @@
+FROM node:16-alpine
```
Comment (cryi) [inline Dockerfile:1]: Full path to container repository.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+RUN apk add python3 make gcc musl-dev pkgconfig libusb-dev g++ linux-headers eudev-dev
```
Comment (cryi) [inline Dockerfile:3]: What are we compiling?

Comment (cryi) [reply Dockerfile:3]: Should be removed altogether

File: .github/workflows/build_container.yml [hunk 3]
```diff
@@ -0,0 +1,44 @@
+name: publish
+
+on:
+  push:
```
Comment (cryi) [inline .github/workflows/build_container.yml:4]: This should be ok to run only on `release`

**Review by cryi — CHANGES_REQUESTED** [review]

File: Dockerfile (hunk 2 above)
Comment (nicolasochem) [reply Dockerfile:3]: all of these are needed for npm install of all breadcrumbs dependencies. the container does not build if you remove any of these packages.

File: .github/workflows/build_container.yml (hunk 3 above)
Comment (nicolasochem) [reply .github/workflows/build_container.yml:4]: why?

Comment (nicolasochem) [reply .github/workflows/build_container.yml:4]: sometimes I need to deploy a hotfix from a PR or an unreleased fix. It's not practical to wait for the fix, besides GHCR is free for open source projects so it's simple and cheap to build everything all the time.

Comment (cryi) [reply .github/workflows/build_container.yml:4]: Technically it is ok but it should be clear that we do not provide support for such images. And you can always build your hotfix yourself and ling your mirror.

**Comment by cryi** [comment]
@nicolasochem Please make sure you are ok with SSPL (#72) It should be merged before this commit.

File: Dockerfile (hunk 2 above)
Comment (nicolasochem) [reply Dockerfile:3]: ok, are you going to move it to optional?

**Comment by kalouo** [comment]
Hi @nicolasochem. I have made dependencies related to Ledger and Parcel optional. The image appears to be building without the extra compilation steps.  Can you check on your end? 

P

**Comment by kalouo** [comment]
Also – could you please make notes in the wiki for how to build the image and run breadcrumbs with Docker? 

Thanks

**Comment by kalouo** [comment]
FYI: 

If a dependency can be used, but you would like npm to proceed if it cannot be found or fails to install, then you may put it in the optionalDependencies object. This is a map of package name to version or url, just like the dependencies object. The difference is that build failures do not cause installation to fail. Running npm install --omit=optional will prevent these dependencies from being installed.

**Comment by nicolasochem** [comment]
Thanks, but if ledger is a supported use case, why not build it always? It's possible to use a ledger inside of a docker container (in privileged mode and by mounting /dev/usb)

What's the value add for the user to make these optional? It could be confusing.

Conversely, what's the problem if the Docker container has more packages? Then, everything you can do on breadcrumbs, you can also do inside docker.

Also, as an added benefit, if something breaks in the future with building with ledger support, we will know straight away because the CI will fail.

In my view, we either have direct ledger support or we don't. if we do, let's have it supported in all build artifacts by default.

**Comment by kalouo** [comment]
Sure @nicolasochem – have reverted the commit. 

Can you clarify if Ledger is already supported here? Or if we need to test it?

**Comment by nicolasochem** [comment]
I didn't test ledger in docker, no.

**Pull request closed without merging**